### PR TITLE
feature/#1099 added check instance types in equal functions

### LINKED ISSUE #1099: When comparing Scenario to other, check that other isinstance Scenario
**Description**
When building a frontend with taipy, I often have a `selected_scenario` state variable which is `Optional[Scenario]`. Now, when I use an expression like `selected_scenario in tp.get_scenarios()`, I have to prepend a `isinstance(selected_scenario, Scenario) and` to avoid an exception.

Example as follows:

```python
from taipy import Config, Scenario
import taipy as tp
from typing import Optional

x_cfg = Config.configure_data_node(id="x")
y_cfg = Config.configure_data_node(id="y")

def square_fn(x: int) -> int: x**2
square_task_cfg = Config.configure_task(id="square", function=square_fn, input=[x_cfg], output=[y_cfg])

scenario_cfg = Config.configure_scenario(id="scenario", task_configs=[square_task_cfg])

if __name__ == "__main__":
    tp.Core().run()
    scenario = tp.create_scenario(scenario_cfg)

    # __eq__
    selected_scenario: Optional[Scenario] = None
    try:
        selected_scenario == scenario
    except AttributeError as ae:
        print(ae)  # 'NoneType' object has no attribute 'id'

    # in, indirectly calls __eq__
    selected_scenario: Optional[Scenario] = None
    try:
        selected_scenario in tp.get_scenarios()
    except AttributeError as ae:
        print(ae)  # 'NoneType' object has no attribute 'id'
```

Would it be appropriate to add a check in `Scenario.__eq__` for this? Same question for the other entities too

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented

### PR DESCRIPTION
**Pull request by toan-quach**
#1099

Changes:
- Add a check in Entities's `__eq__` for type instance

### CODE REVIEW

File: taipy/core/cycle/cycle.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def __getattr__(self, attribute_name):
         raise AttributeError(f"{attribute_name} is not an attribute of cycle {self.id}")
 
     def __eq__(self, other):
-        return self.id == other.id
+        return isinstance(other, Cycle) and self.id == other.id
```
Comment (trgiangdo) [inline taipy/core/cycle/cycle.py:160]: should we raise an error instead of just False if the `other` is not a Cycle?

**Review by trgiangdo — DISMISSED** [review]
LGTM. I just have a question

File: taipy/core/cycle/cycle.py (hunk 1 above)
Comment (jrobinAV) [reply taipy/core/cycle/cycle.py:160]: I would not raise an error. This code looks correct to me.

File: tests/core/scenario/test_scenario.py [hunk 2]
```diff
@@ -32,6 +32,21 @@
...
+    scenario_manager._set(scenario)
+
+    # To test if instance is same type
+    task = Task("task", {}, print, [], [], scenario_id)
+
+    scenario_2 = scenario_manager._get(scenario_id)
+    assert scenario == scenario_2
+    assert scenario != scenario_id
+    assert scenario != task
+
```
Comment (jrobinAV) [inline tests/core/scenario/test_scenario.py:48]:
```suggestion
We should add a test closer to our user's case using optionals.

```suggestion
    scenario = tp.create_scenario(scenario_cfg)

    # __eq__
    selected_scenario: Optional[Scenario] = None
    try:
        selected_scenario == scenario
    except AttributeError as ae:
        print(ae)  # 'NoneType' object has no attribute 'id'

    # in, indirectly calls __eq__
    selected_scenario: Optional[Scenario] = None
    try:
        selected_scenario in tp.get_scenarios()
    except AttributeError as ae:
        print(ae)  # 'NoneType' object has no attribute 'id'
```

What do you think?

**Review by jrobinAV — COMMENTED** [review]
I just have a proposal to enhance the test suite.

File: taipy/core/cycle/cycle.py (hunk 1 above)
Comment (toan-quach) [reply taipy/core/cycle/cycle.py:160]: same, I wouldn't raise an error also, the purpose is to check if this `other` value is equal to the Cycle, if it's not, then the right thing to do is return False. We don't need to raise any errors for it

File: tests/core/scenario/test_scenario.py (hunk 2 above)
Comment (toan-quach) [reply tests/core/scenario/test_scenario.py:48]: I think all the cases have been covered (None, same Scenario, different type instance). Not sure if it's needed :/

Comment (jrobinAV) [reply tests/core/scenario/test_scenario.py:48]: I don't see where we check for None values.

**Review by jrobinAV — DISMISSED** [review]
Minor test improvement about asserting `scenario != None` values<

**Comment by toan-quach** [comment]
@jrobinAV 
I actually added but pre-commit doesn't allow it and `type: ignore` doesn't help also :/

**Review by jrobinAV — APPROVED** [review]

**Pull request merged**